src/TopDownProteomics/ProForma/TopPicProformaParser.cs

```diff
@@ -0,0 +1,186 @@
+            {
+                char c = sequence[i];
+                if (c == '[')
+                    inBraket = true;
+                else if (c == ']')
+                    inBraket = false;
+                else if (char.IsUpper(c) && !inBraket)
+                {
+                    indexLoopup[i] = index++;
+                }
+            }
+            return indexLoopup;
+        }
+
+        private Tuple<List<ProFormaDescriptor>, List<ProFormaDescriptor>, List<ProFormaTag>> FindPTMs(string sequence)
+        {
+            var indexLoopup = GetIndexLoopup(sequence);
+
+            List<ProFormaDescriptor> N_terms = new List<ProFormaDescriptor>();
+            List<ProFormaDescriptor> C_terms = new List<ProFormaDescriptor>();
+            List<ProFormaTag> Tags = new List<ProFormaTag>();
+
+            foreach (Match match in _modRx.Matches(sequence))
+            {
+                var startIndex = indexLoopup[match.Groups[1].Index];
+                var ptms = match.Groups[2].Captures;
+
+                if (ptms.Count > 1)
+                    throw new Exception("multiple mods are not currently accepeted");
+
+                if (startIndex == 0 && match.Groups[1].Length == 1)  // check for ambiguoous mods that include the start -> just make tags
+                {
+                    N_terms = ParsePTMs(ptms);
+                }
+                else if (startIndex == indexLoopup.Max(x => x.Value))
+                {
+                    C_terms = ParsePTMs(ptms);
+                }
+                else if (match.Groups[1].Length > 1)
+                {
+                    var EndIndex = startIndex + match.Groups[1].Length - 1;
+                    Tags.Add(new ProFormaTag(startIndex, EndIndex, ParsePTMs(ptms)));
+                }
+                else
+                    Tags.Add(new ProFormaTag(startIndex, ParsePTMs(ptms)));
+            }
+            return new Tuple<List<ProFormaDescriptor>, List<ProFormaDescriptor>, List<ProFormaTag>>(N_terms, C_terms, Tags);
+        }
+
+        private List<ProFormaDescriptor> ParsePTMs(CaptureCollection ptms)
+        {
+            var proformaList = new List<ProFormaDescriptor>();
+
+            foreach (var ptm in ptms)
+                proformaList.Add(ParsePTMstring(ptm.ToString()));
+
+            return proformaList;
+        }
+
+        private ProFormaDescriptor ParsePTMstring(string ptmstring)
+        {
+            //strip []
+            ptmstring = ptmstring.Substring(1, ptmstring.Length - 2);
+            var numberMatch = _numberRx.Match(ptmstring);
+
+            if (numberMatch.Success)
+                return new ProFormaDescriptor(ProFormaKey.Mass, numberMatch.Value);
+
+            // Find and throw exception if there is a *
+            if (ptmstring.Contains('*'))
+                throw new Exception("multiple mods are not currently supported");
+
+            if (_modLookup is not null && _modLookup.ContainsKey(ptmstring))
```

Comment:
could do `if (_modLookup?.ContainsKey(ptmstring) == true)` if you like that better